tests/README.md

```diff
@@ -0,0 +1,1510 @@
+	}
+
+
+	/**
+	 * @Test
+	 * @small
+	 */
+	public function testReadWhichShouldSucceed() {
+		//Arrange
+		$fileContent = 'Test file content.';
+		$this->filesystemMock->shouldReceive('read')
+			->once()
+			->andReturn($fileContent);
+
+		//Act
+		$actualContent = $this->subject->read('/path/to/your/file');
+
+		//Assert
+		$this->assertSame($fileContent, $actualContent);
+	}
+}
+```
+
+<a name="fat-legacy-class-with-parent"></a>
+#### Fat legacy class with parent
+Fat legacy classes especially in ILIAS extend each other to do some work.
+For example the ilObject2 has a create method which depends on the create method of the ilObject.
+The ilObject create method writes the data to the database which is not desirable in unit tests of ilObject2.
+Therefore ilObject should be replaced with a stub and ilObject2 should operate on that stub which has
+proper expectation in place to test the behaviour.
+
+This requires an autoload "hack" which loads the base class mock (ilObject) before the partial mock
+ilObject2 is created. Important is that all expectation are set on the ilObject mock before the
+ilObject2 instance is created otherwise the expectations have no effect.
+
+```php
+<?php
+require_once './libs/composer/vendor/autoload.php';
+
+use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Class ilObject2Test
+ *
+ * @author  Nicolas Schäfli <ns@studer-raimann.ch>
+ *
+ * @runTestsInSeparateProcesses
+ * @preserveGlobalState    disabled
+ * @backupGlobals          disabled
+ * @backupStaticAttributes disabled
+ */
+class ilObject2Test extends TestCase {
+
+	use MockeryPHPUnitIntegration;
+	/**
+	 * @var ilObject2 $subject
+	 */
+	private $subject;
+	/**
+	 * @var ilObject | \Mockery\MockInterface $subjectParent
+	 */
+	private $subjectParent;
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function testCreateWithoutCloneModeWhichShouldSucceed() {
+
+		//expectations must be set before the child class is loaded
+		$expectedId = 5;
+		$this->subjectParent = Mockery::mock('overload:' . ilObject::class);
+		$this->subjectParent->shouldReceive('create')
+			->once()
+			->andReturn($expectedId);
+		$this->subject = Mockery::mock(ilObject2::class . '[]', []);
+
+		$result = $this->subject->create();
+
+		$this->assertSame($expectedId, $result);
+	}
+
+	//test other cases ...
+}
+```   
+In cases as shown above, each test has to be run in a separate PHP process because of the specially loaded classes. This can be done with the *@runTestsInSeparateProcesses* annotation.
+
+<a name="partial-mocks"></a>
+##### Partial mocks
+On the line with the suspicious looking string concatenation of the class name with
+empty brackets:
+
+```
+$this->subject = Mockery::mock(ilObject2::class . '[]', []);
+```
+
+This is directly creating a partial mock without any mocked methods. That is
+useful to test abstract classes without an actual implementation. The second parameter
```

Comment:
I don't understand that.

Reply:
A partial mock holds an instance of the class which is mocked and delegates the not mocked methods to the real implementation. 

We are not able to create instances of abstract classes to test them. However mockery has some internal magic which allows use to create a mock object of the abstract class and to delegate all methods to it. Which allows us to test them like normal.

Reply:
Thx. That could go somewhere in the actual text =)